File: packages/mobile/android/core/src/main/kotlin/com/doublebind/core/DatabaseModule.kt
/**
 * React Native native module that provides database utilities for the mobile app.
 *
 * With the migration to op-sqlite, this module only needs to provide the database path
 * since op-sqlite handles all database operations via JSI.
 */
package com.doublebind.core

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import java.io.File

/**
 * React Native module providing database utilities.
 * Provides the database path for op-sqlite to use.
 */
class DatabaseModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    /**
     * Module name exposed to JavaScript via NativeModules.
     */
    override fun getName(): String = "DatabaseModule"

    /**
     * Get the default database path for the app.
     * Returns the path to the app's databases directory where the database should be stored.
     *
     * @param promise Resolves with the database path, rejects on error
     */
    @ReactMethod
    fun getDatabasePath(promise: Promise) {
        try {
            val context = reactApplicationContext
            val databasesDir = context.getDatabasePath("dummy").parentFile
                ?: throw IllegalStateException("Failed to get databases directory")

            // Ensure the databases directory exists
            if (!databasesDir.exists()) {
                databasesDir.mkdirs()
            }

            val dbPath = File(databasesDir, "double-bind.db").absolutePath
            promise.resolve(dbPath)
        } catch (e: Exception) {
            promise.reject("ERR_GET_PATH", "Failed to get database path: ${e.message}", e)
        }
    }

    /**
     * Ensure the database directory exists.
     * Creates the parent directory for the database file if it doesn't exist.
     *
     * @param path Path to the database file
     * @param promise Resolves on success, rejects with error on failure
     */
    @ReactMethod
    fun ensureDatabaseDirectory(path: String, promise: Promise) {
        try {
            val file = File(path)
            val parentDir = file.parentFile
                ?: throw IllegalArgumentException("Invalid database path: $path")

            if (!parentDir.exists()) {
                if (!parentDir.mkdirs()) {
                    throw IllegalStateException("Failed to create directory: ${parentDir.absolutePath}")
                }
            }

            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("ERR_CREATE_DIR", "Failed to create database directory: ${e.message}", e)
        }
    }
}

File: packages/mobile/ios/DoubleBindCore/Sources/DatabaseModule.swift
//
//  DatabaseModule.swift
//  DoubleBindCore
//
//  React Native native module that provides database utilities for the mobile app.
//  With the migration to op-sqlite, this module only needs to provide the database path
//  since op-sqlite handles all database operations via JSI.
//

import Foundation
import React

/// React Native native module for database utilities.
/// Provides the database path for op-sqlite to use.
@objc(DatabaseModule)
final class DatabaseModule: NSObject {

    /// Get the default database path for the app.
    /// Returns the path to the app's Documents directory where the database should be stored.
    ///
    /// - Parameters:
    ///   - resolve: Promise resolve callback returning the database path
    ///   - reject: Promise reject callback
    @objc
    func getDatabasePath(
        _ resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        // Get the app's Documents directory
        guard let documentsPath = FileManager.default.urls(
            for: .documentDirectory,
            in: .userDomainMask
        ).first?.path else {
            reject(
                "ERR_NO_DOCUMENTS_DIR",
                "Failed to get documents directory path",
                nil
            )
            return
        }

        // Return the path to the database file
        let dbPath = (documentsPath as NSString).appendingPathComponent("double-bind.db")
        resolve(dbPath)
    }

    /// Ensure the database directory exists.
    /// Creates the parent directory for the database file if it doesn't exist.
    ///
    /// - Parameters:
    ///   - path: Path to the database file
    ///   - resolve: Promise resolve callback
    ///   - reject: Promise reject callback
    @objc
    func ensureDatabaseDirectory(
        _ path: String,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        let directoryPath = (path as NSString).deletingLastPathComponent

        do {
            try FileManager.default.createDirectory(
                atPath: directoryPath,
                withIntermediateDirectories: true,
                attributes: nil
            )
            resolve(nil)
        } catch {
            reject(
                "ERR_CREATE_DIR",
                "Failed to create database directory at \(directoryPath): \(error.localizedDescription)",
                error
            )
        }
    }
}
